src/tool/abi/abi_writer.h

```diff
@@ -0,0 +1,260 @@
+#pragma once
+
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "common.h"
+#include "generic_arg_stack.h"
+#include "meta_reader.h"
+#include "namespace_iterator.h"
+#include "text_writer.h"
+
+struct console_writer : xlang::text::writer_base<console_writer> {};
+
+template <typename Int>
+struct format_hex
+{
+    static_assert(std::is_unsigned_v<Int>);
+    Int value;
+};
+template <typename Int>
+format_hex(Int) -> format_hex<Int>;
+
+struct indent { std::size_t additional_indentation = 0; };
+
+struct writer : xlang::text::writer_base<writer>
+{
+    writer(
+        std::string_view ns,
+        abi_configuration const& config,
+        xlang::meta::reader::cache const& cache,
+        xlang::meta::reader::cache::namespace_members const& members) :
+        m_namespace(ns),
+        m_config(config),
+        m_cache(cache),
+        m_members(members)
+    {
+        initialize_dependencies();
+    }
+
+    using writer_base::write;
+
+    abi_configuration const& config() const noexcept
+    {
+        return m_config;
+    }
+
+    void write_code(std::string_view value)
+    {
+        xlang::text::bind_list("::", namespace_range{ value })(*this);
+    }
+
+    static void write_uppercase(writer& w, std::string_view str)
+    {
+        for (auto ch : str)
+        {
+            w.write(static_cast<char>(::toupper(ch)));
+        }
+    }
+
+    static void write_lowercase(writer& w, std::string_view str)
+    {
+        for (auto ch : str)
+        {
+            w.write(static_cast<char>(::tolower(ch)));
+        }
+    }
+
+    template <typename Int>
+    void write(format_hex<Int> val)
+    {
+        write_value(val.value);
+    }
+
+    void write(indent value)
+    {
+        for (std::size_t i = 0; i < (m_indentation + value.additional_indentation); ++i)

```

Comment:
>m_indentation [](start = 37, length = 13)

I think Harry is refactoring to move indent guard into the base library

Reply:
Cool 👍🏼

---
In reply to: [227979165](https://github.com/Microsoft/xlang/pull/51#discussion_r227979165) [](ancestors = 227979165)